<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>R Breeding Simulator • breedSimulatR</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="R Breeding Simulator">
<meta property="og:description" content="R package providing classes and functions to
    simulate breeding schemes.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">breedSimulatR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.2.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="news/index.html">News</a>
</li>
<li>
  <a href="https://github.com/ut-biomet/breedSimulatR/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div id="breedsimulatr" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#breedsimulatr" class="anchor"></a>breedSimulatR</h1></div>
<!-- badges: start -->

<p>R package providing classes and functions to simulate breeding schemes.</p>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<p>:octocat: You can install <code>breedSimulatR</code> from <a href="https://github.com/">GitHub</a> with:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html">require</a></span><span class="op">(</span><span class="st"><a href="https://devtools.r-lib.org/">"devtools"</a></span><span class="op">)</span><span class="op">)</span> <span class="op">{</span>
    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span>
<span class="op">}</span>
<span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://devtools.r-lib.org//reference/remote-reexports.html">install_github</a></span><span class="op">(</span><span class="st">"ut-biomet/breedSimulatR"</span><span class="op">)</span></code></pre></div>
<p>You can check the installation with these lines:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/ut-biomet/breedSimulatR">breedSimulatR</a></span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/help.html">help</a></span><span class="op">(</span>package <span class="op">=</span> <span class="va">breedSimulatR</span><span class="op">)</span></code></pre></div>
</div>
<div id="example" class="section level2">
<h2 class="hasAnchor">
<a href="#example" class="anchor"></a>Example</h2>
<p>This is a basic example which shows how to use the package.</p>
<p>The package contains some example data that we will use for this example. These data are stored in the variable <code>exampleData</code>.</p>
<p><code>exampleData</code> is a list containing 3 elements:</p>
<ul>
<li>
<code>exampleData$genotypes</code>: <code>data.frame</code> containing the genotypic data encoded in allele doses of 100 fictitious individuals with 3333 SNP markers. These individuals have 10 chromosomes of length 10^6 bases pairs.</li>
<li>
<code>exampleData$snpCoord</code>: <code>data.frame</code> containing the coordinates of the 3333 individuals’ markers. This data.frame contains 3 columns: <code>chr</code>, <code>pos</code> and <code>SNPid</code>.</li>
<li>
<code>exampleData$snpEffects</code>: <code>numeric</code> vector containing the “true” effects of the 3333 individuals’ markers about a fictitious quantitative trait based on an additive architecture.</li>
</ul>
<div id="initialization" class="section level4">
<h4 class="hasAnchor">
<a href="#initialization" class="anchor"></a>initialization</h4>
<p>First we must load the package:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/ut-biomet/breedSimulatR">breedSimulatR</a></span><span class="op">)</span></code></pre></div>
</div>
<div id="specie-specification" class="section level4">
<h4 class="hasAnchor">
<a href="#specie-specification" class="anchor"></a>Specie specification</h4>
<p>Let’s specify the specie:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># create specie object</span>
<span class="va">specie_statEx</span> <span class="op">&lt;-</span> <span class="va"><a href="reference/specie.html">specie</a></span><span class="op">$</span><span class="fu">new</span><span class="op">(</span>specName <span class="op">=</span> <span class="st">"Statisticae exempli"</span>, nChr <span class="op">=</span> <span class="fl">10</span>, lchr <span class="op">=</span> <span class="fl">1000000</span>, 
    lchrCm <span class="op">=</span> <span class="fl">100</span>, ploidy <span class="op">=</span> <span class="fl">2</span><span class="op">)</span>
<span class="co">#&gt; A new species has emerged: Statisticae exempli !</span></code></pre></div>
</div>
<div id="snp-specification" class="section level4">
<h4 class="hasAnchor">
<a href="#snp-specification" class="anchor"></a>SNP specification</h4>
<p>We must specify the information about the positions of the genotypic markers used in the simulation.</p>
<p>Let’s load these information (stored in <code>exampleData$snpCoord</code>) and create the <code>SNPinfo</code> object.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># data preview</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">exampleData</span><span class="op">$</span><span class="va">snpCoord</span><span class="op">)</span>
<span class="co">#&gt;      chr physPos linkMapPos    SNPid</span>
<span class="co">#&gt; 2  Chr01  937638   86.81546 snp03760</span>
<span class="co">#&gt; 6  Chr01  654763   56.51842 snp02674</span>
<span class="co">#&gt; 10 Chr01  181658   30.62157 snp00721</span>
<span class="co">#&gt; 18 Chr01  230126   35.47120 snp00948</span>
<span class="co">#&gt; 26 Chr01  420637   46.98455 snp01620</span>
<span class="co">#&gt; 29 Chr01  467620   48.80905 snp01790</span></code></pre></div>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># create SNPinfo object</span>
<span class="va">SNPs</span> <span class="op">&lt;-</span> <span class="va"><a href="reference/SNPinfo.html">SNPinfo</a></span><span class="op">$</span><span class="fu">new</span><span class="op">(</span>SNPcoord <span class="op">=</span> <span class="va">exampleData</span><span class="op">$</span><span class="va">snpCoord</span>, specie <span class="op">=</span> <span class="va">specie_statEx</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">SNPs</span><span class="op">)</span>
<span class="co">#&gt; specie: Statisticae exempli</span>
<span class="co">#&gt; 3333 Markers on 10 chromosomes :</span>
<span class="co">#&gt; Chr01 Chr02 Chr03 Chr04 Chr05 Chr06 Chr07 Chr08 Chr09 Chr10 </span>
<span class="co">#&gt;   415   247   425   322   381   269   238   355   342   339 </span>
<span class="co">#&gt; SNPcoord:</span>
<span class="co">#&gt;        chr physPos linkMapPos    SNPid</span>
<span class="co">#&gt; 338  Chr01    2068  0.4911687 snp00006</span>
<span class="co">#&gt; 1450 Chr01    2708  0.6423784 snp00009</span>
<span class="co">#&gt; 1391 Chr01    2782  0.6598378 snp00011</span>
<span class="co">#&gt; 2475 Chr01    4159  0.9838113 snp00018</span>
<span class="co">#&gt; 1553 Chr01    6917  1.6275084 snp00026</span>
<span class="co">#&gt; 2365 Chr01    7814  1.8353769 snp00031</span>
<span class="co">#&gt;  [ reached 'max' / getOption("max.print") -- omitted 3327 rows ]</span></code></pre></div>
</div>
<div id="population-initialization" class="section level4">
<h4 class="hasAnchor">
<a href="#population-initialization" class="anchor"></a>Population initialization</h4>
<p>We can now generate an initial population from genotypic data.</p>
<p>Let’s load the genotypic information (stored in <code>exampleData$genotypes</code>) and create the <code>population</code> object:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># data preview</span>
<span class="va">exampleData</span><span class="op">$</span><span class="va">genotypes</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span>
<span class="co">#&gt;          snp00006 snp00009 snp00011 snp00018 snp00026</span>
<span class="co">#&gt; Coll0001        2        2        2        0        2</span>
<span class="co">#&gt; Coll0002        0        2        2        2        0</span>
<span class="co">#&gt; Coll0003        2        2        2        0        2</span></code></pre></div>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># create population object</span>
<span class="va">initPop</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/createPop.html">createPop</a></span><span class="op">(</span>geno <span class="op">=</span> <span class="va">exampleData</span><span class="op">$</span><span class="va">genotypes</span>, SNPinfo <span class="op">=</span> <span class="va">SNPs</span>, popName <span class="op">=</span> <span class="st">"Initial population"</span><span class="op">)</span></code></pre></div>
</div>
<div id="traits-and-phenotyping-initialization" class="section level4">
<h4 class="hasAnchor">
<a href="#traits-and-phenotyping-initialization" class="anchor"></a>Traits and phenotyping initialization</h4>
<p>Let’s create 2 independent phenotypic traits that can be phenotyped.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">nQtn</span> <span class="op">&lt;-</span> <span class="fl">1000</span>

<span class="va">qtn</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">initPop</span><span class="op">$</span><span class="va">maf</span> <span class="op">&gt;</span> <span class="fl">0.1</span><span class="op">)</span>, <span class="va">nQtn</span><span class="op">)</span>
<span class="va">weight</span> <span class="op">&lt;-</span> <span class="va"><a href="reference/trait.html">trait</a></span><span class="op">$</span><span class="fu">new</span><span class="op">(</span>name <span class="op">=</span> <span class="st">"Weight"</span>, qtn <span class="op">=</span> <span class="va">qtn</span>, qtnEff <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="va">nQtn</span>, <span class="fl">0</span>, <span class="fl">0.35</span><span class="op">)</span><span class="op">)</span>

<span class="va">qtn</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">initPop</span><span class="op">$</span><span class="va">maf</span> <span class="op">&gt;</span> <span class="fl">0.1</span><span class="op">)</span>, <span class="va">nQtn</span><span class="op">)</span>
<span class="va">height</span> <span class="op">&lt;-</span> <span class="va"><a href="reference/trait.html">trait</a></span><span class="op">$</span><span class="fu">new</span><span class="op">(</span>name <span class="op">=</span> <span class="st">"Height"</span>, qtn <span class="op">=</span> <span class="va">qtn</span>, qtnEff <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="va">nQtn</span>, <span class="fl">0</span>, <span class="fl">0.25</span><span class="op">)</span><span class="op">)</span>

<span class="va">phenolab</span> <span class="op">&lt;-</span> <span class="va"><a href="reference/phenotyper.html">phenotyper</a></span><span class="op">$</span><span class="fu">new</span><span class="op">(</span>name <span class="op">=</span> <span class="st">"Pheno lab"</span>, traits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="va">weight</span>, <span class="va">height</span><span class="op">)</span>, plotCost <span class="op">=</span> <span class="fl">150</span>, 
    mu <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">100</span>, <span class="fl">75</span><span class="op">)</span>, he <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0.4</span>, <span class="fl">0.6</span><span class="op">)</span>, pop <span class="op">=</span> <span class="va">initPop</span><span class="op">)</span>

<span class="va">pheno</span> <span class="op">&lt;-</span> <span class="va">phenolab</span><span class="op">$</span><span class="fu">trial</span><span class="op">(</span>pop <span class="op">=</span> <span class="va">initPop</span>, rep <span class="op">=</span> <span class="fl">4</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">pheno</span><span class="op">$</span><span class="va">data</span><span class="op">)</span>
<span class="co">#&gt;        ind    Weight   Height rep phenotyper</span>
<span class="co">#&gt; 1 Coll0001  92.95343 74.58895   1  Pheno lab</span>
<span class="co">#&gt; 2 Coll0001  95.25723 77.01605   2  Pheno lab</span>
<span class="co">#&gt; 3 Coll0001 106.49726 76.35178   3  Pheno lab</span>
<span class="co">#&gt; 4 Coll0001  89.47586 75.33123   4  Pheno lab</span>
<span class="co">#&gt; 5 Coll0002 105.92757 82.12443   1  Pheno lab</span>
<span class="co">#&gt;  [ reached 'max' / getOption("max.print") -- omitted 1 rows ]</span>
<span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">pheno</span><span class="op">$</span><span class="va">cost</span><span class="op">)</span>
<span class="co">#&gt; [1] 60000</span></code></pre></div>
</div>
<div id="selection-simulation" class="section level4">
<h4 class="hasAnchor">
<a href="#selection-simulation" class="anchor"></a>Selection Simulation</h4>
<p>In order to perform crossing, we must specify which individuals must be mate together. Therefore, we must create functions which generate a crossing table from our population.</p>
<p>For this example, we will use the function <code>selectBV</code>, which returns the names of the best individuals according to their breeding values.</p>
<p>Then, the function <code>randomMate</code> will generate the crossing table.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">exampleData</span><span class="op">$</span><span class="va">snpEffects</span>
<span class="co">#&gt;     snp00006     snp00009     snp00011     snp00018     snp00026     snp00031 </span>
<span class="co">#&gt;  0.051109298  0.080992331  0.183532594  0.028194339  0.144136946 -0.044546216 </span>
<span class="co">#&gt;     snp00035     snp00036     snp00049     snp00052     snp00075     snp00076 </span>
<span class="co">#&gt; -0.065193108  0.043784786  0.152643726 -0.198029104  0.066515171 -0.231082616 </span>
<span class="co">#&gt;     snp00087     snp00101     snp00102     snp00103     snp00111     snp00113 </span>
<span class="co">#&gt;  0.008411049  0.016470385 -0.029577718  0.157991987 -0.073798339  0.045285975 </span>
<span class="co">#&gt;     snp00116     snp00117     snp00130     snp00131     snp00141     snp00144 </span>
<span class="co">#&gt;  0.022954650  0.041486485 -0.087808853  0.135689743  0.077546360  0.086486365 </span>
<span class="co">#&gt;     snp00145 </span>
<span class="co">#&gt;  0.072675256 </span>
<span class="co">#&gt;  [ reached getOption("max.print") -- omitted 3308 entries ]</span>
<span class="op">(</span><span class="va">selectedInds</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/selectBV.html">selectBV</a></span><span class="op">(</span>pop <span class="op">=</span> <span class="va">initPop</span>, SNPeffects <span class="op">=</span> <span class="va">exampleData</span><span class="op">$</span><span class="va">snpEffects</span>, n <span class="op">=</span> <span class="fl">10</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt;  [1] "Coll0068" "Coll0008" "Coll0074" "Coll0016" "Coll0020" "Coll0045"</span>
<span class="co">#&gt;  [7] "Coll0079" "Coll0002" "Coll0046" "Coll0097"</span>

<span class="op">(</span><span class="va">crossTable</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/randomMate.html">randomMate</a></span><span class="op">(</span>inds <span class="op">=</span> <span class="va">selectedInds</span>, n <span class="op">=</span> <span class="fl">120</span>, names <span class="op">=</span> <span class="st">"generation_1"</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt;       ind1     ind2 n            names</span>
<span class="co">#&gt; 1 Coll0079 Coll0008 1 generation_1-001</span>
<span class="co">#&gt; 2 Coll0008 Coll0045 1 generation_1-002</span>
<span class="co">#&gt; 3 Coll0074 Coll0097 1 generation_1-003</span>
<span class="co">#&gt; 4 Coll0045 Coll0097 1 generation_1-004</span>
<span class="co">#&gt; 5 Coll0068 Coll0068 1 generation_1-005</span>
<span class="co">#&gt; 6 Coll0045 Coll0046 1 generation_1-006</span>
<span class="co">#&gt;  [ reached 'max' / getOption("max.print") -- omitted 114 rows ]</span></code></pre></div>
<p>We can now generate the offspring:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">newPop</span> <span class="op">&lt;-</span> <span class="va"><a href="reference/population.html">population</a></span><span class="op">$</span><span class="fu">new</span><span class="op">(</span>name <span class="op">=</span> <span class="st">"1st offspring"</span>, inds <span class="op">=</span> <span class="fu"><a href="reference/makeCrosses.html">makeCrosses</a></span><span class="op">(</span>crosses <span class="op">=</span> <span class="va">crossTable</span>, 
    pop <span class="op">=</span> <span class="va">initPop</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">newPop</span>
<span class="co">#&gt; Population: 1st offspring</span>
<span class="co">#&gt; Species: Statisticae exempli</span>
<span class="co">#&gt; Number of individuals: 120</span></code></pre></div>
<p>This process can be included in loops in order to simulate several generations.</p>
</div>
</div>
<div id="issues" class="section level2">
<h2 class="hasAnchor">
<a href="#issues" class="anchor"></a>Issues</h2>
<p>When encountering a problem with the package or if you have questions, please report issues on GitHub <a href="https://github.com/ut-biomet/breedSimulatR/issues">here</a>.</p>
<p>I will do my best to help you as soon as possible.</p>
</div>
<div id="contributing" class="section level2">
<h2 class="hasAnchor">
<a href="#contributing" class="anchor"></a>Contributing</h2>
<p>You can contribute in various ways:</p>
<ul>
<li>report an <a href="https://github.com/ut-biomet/breedSimulatR/issues">issue</a> (online, see the above section)</li>
<li>suggest improvements (in the same way as issues)</li>
<li>propose a pull request (after creating a new branch)</li>
</ul>
<p>When editing the content of this package, please run the following commands before asking a pull request:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://devtools.r-lib.org//reference/document.html">document</a></span><span class="op">(</span><span class="op">)</span>
<span class="va">pkg</span> <span class="op">&lt;-</span> <span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://devtools.r-lib.org//reference/build.html">build</a></span><span class="op">(</span><span class="op">)</span>
<span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://devtools.r-lib.org//reference/check.html">check_built</a></span><span class="op">(</span><span class="va">pkg</span><span class="op">)</span></code></pre></div>
</div>
<div id="citation" class="section level2">
<h2 class="hasAnchor">
<a href="#citation" class="anchor"></a>Citation</h2>
<p>Please cite this package when using it for your projects:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/citation.html">citation</a></span><span class="op">(</span><span class="st">"breedSimulatR"</span><span class="op">)</span></code></pre></div>
<p>See also <code><a href="https://rdrr.io/r/utils/citation.html">citation()</a></code> for citing R itself.</p>
</div>
<div id="acknowledgments" class="section level2">
<h2 class="hasAnchor">
<a href="#acknowledgments" class="anchor"></a>Acknowledgments</h2>
<p>Thanks to <a href="https://github.com/KosukeHamazaki">Kosuke Hamazaki</a> for his feedbacks.</p>
</div>
<div id="references" class="section level2">
<h2 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h2>
<p><code>breedSimulatR</code> is written in <strong>R</strong>:</p>
<ul>
<li>
R Core Team (2020).<em>R: A Language and Environment for Statistical Computing</em>.R Foundation for Statistical Computing, Vienna, Austria.<a href="https://www.R-project.org/">https://www.R-project.org/</a>.
</li>
</ul>
<p><code>breedSimulatR</code> package or its development required the following R packages:</p>
<ul>
<li>
<b> plotly </b>
<ul>
<li>
Sievert C (2020).<em>Interactive Web-Based Data Visualization with R, plotly, and shiny</em>.Chapman and Hall/CRC.ISBN 9781138331457, <a href="https://plotly-r.com">https://plotly-r.com</a>.
</li>
</ul>
</li>
<li>
<b> R6 </b>
<ul>
<li>
Chang W (2020).<em>R6: Encapsulated Classes with Reference Semantics</em>.R package version 2.5.0, <a href="https://CRAN.R-project.org/package=R6">https://CRAN.R-project.org/package=R6</a>.
</li>
</ul>
</li>
<li>
<b> covr </b>
<ul>
<li>
Hester J (2020).<em>covr: Test Coverage for Packages</em>.R package version 3.5.1, <a href="https://CRAN.R-project.org/package=covr">https://CRAN.R-project.org/package=covr</a>.
</li>
</ul>
</li>
<li>
<b> devtools </b>
<ul>
<li>
Wickham H, Hester J, Chang W (2020).<em>devtools: Tools to Make Developing R Packages Easier</em>.R package version 2.3.1, <a href="https://CRAN.R-project.org/package=devtools">https://CRAN.R-project.org/package=devtools</a>.
</li>
</ul>
</li>
<li>
<b> knitr </b>
<ul>
<li>
Xie Y (2020).<em>knitr: A General-Purpose Package for Dynamic Report Generation in R</em>.R package version 1.30, <a href="https://yihui.org/knitr/">https://yihui.org/knitr/</a>.
</li>
<li>
Xie Y (2015).<em>Dynamic Documents with R and knitr</em>, 2nd edition.Chapman and Hall/CRC, Boca Raton, Florida.ISBN 978-1498716963, <a href="https://yihui.org/knitr/">https://yihui.org/knitr/</a>.
</li>
<li>
Xie Y (2014).“knitr: A Comprehensive Tool for Reproducible Research in R.”In Stodden V, Leisch F, Peng RD (eds.), <em>Implementing Reproducible Computational Research</em>.Chapman and Hall/CRC.ISBN 978-1466561595, <a href="http://www.crcpress.com/product/isbn/9781466561595">http://www.crcpress.com/product/isbn/9781466561595</a>.
</li>
</ul>
</li>
<li>
<b> pkgdown </b>
<ul>
<li>
Wickham H, Hesselberth J (2020).<em>pkgdown: Make Static HTML Documentation for a Package</em>.R package version 1.6.1, <a href="https://CRAN.R-project.org/package=pkgdown">https://CRAN.R-project.org/package=pkgdown</a>.
</li>
</ul>
</li>
<li>
<b> rmarkdown </b>
<ul>
<li>
Allaire J, Xie Y, McPherson J, Luraschi J, Ushey K, Atkins A, Wickham H, Cheng J, Chang W, Iannone R (2020).<em>rmarkdown: Dynamic Documents for R</em>.R package version 2.6, <a href="https://github.com/rstudio/rmarkdown">https://github.com/rstudio/rmarkdown</a>.
</li>
<li>
Xie Y, Allaire J, Grolemund G (2018).<em>R Markdown: The Definitive Guide</em>.Chapman and Hall/CRC, Boca Raton, Florida.ISBN 9781138359338, <a href="https://bookdown.org/yihui/rmarkdown">https://bookdown.org/yihui/rmarkdown</a>.
</li>
<li>
Xie Y, Dervieux C, Riederer E (2020).<em>R Markdown Cookbook</em>.Chapman and Hall/CRC, Boca Raton, Florida.ISBN 9780367563837, <a href="https://bookdown.org/yihui/rmarkdown-cookbook">https://bookdown.org/yihui/rmarkdown-cookbook</a>.
</li>
</ul>
</li>
<li>
<b> roxygen2 </b>
<ul>
<li>
Wickham H, Danenberg P, Csárdi G, Eugster M (2020).<em>roxygen2: In-Line Documentation for R</em>.R package version 7.1.1, <a href="https://CRAN.R-project.org/package=roxygen2">https://CRAN.R-project.org/package=roxygen2</a>.
</li>
</ul>
</li>
<li>
<b> spelling </b>
<ul>
<li>
Ooms J, Hester J (2020).<em>spelling: Tools for Spell Checking in R</em>.R package version 2.2, <a href="https://CRAN.R-project.org/package=spelling">https://CRAN.R-project.org/package=spelling</a>.
</li>
</ul>
</li>
<li>
<b> testthat </b>
<ul>
<li>
Wickham H (2011).“testthat: Get Started with Testing.”<em>The R Journal</em>, <b>3</b>, 5–10.<a href="https://journal.r-project.org/archive/2011-1/RJournal_2011-1_Wickham.pdf">https://journal.r-project.org/archive/2011-1/RJournal_2011-1_Wickham.pdf</a>.
</li>
</ul>
</li>
</ul>
<p>Example data were generated using the serious game “PlantBreedGame” available on GitHub <a href="https://github.com/timflutre/PlantBreedGame" class="uri">https://github.com/timflutre/PlantBreedGame</a></p>
<ul>
<li>
<b>PlantBreedGame</b>
<ul>
<li>
Flutre T, Diot J, and David J (2019). <em>PlantBreedGame</em>: A Serious Game that Puts Students in the Breeder’s Seat. <em>Crop Science.</em> <a href="https://dl.sciencesocieties.org/publications/cs/abstracts/59/4/1374">DOI 10.2135/cropsci2019.03.0183le</a>
</li>
</ul>
</li>
</ul>
</div>
<div id="license-and-copyright-copyright" class="section level2">
<h2 class="hasAnchor">
<a href="#license-and-copyright-copyright" class="anchor"></a>License and copyright <span class="emoji" data-emoji="copyright">©️</span>
</h2>
<p>The <code>breedSimulatR</code> package as a whole is licensed under the MIT. See the <a href="LICENSE">LICENSE</a> file for more details.</p>
<p><span class="emoji" data-emoji="copyright">©️</span> The copyright holder is <a href="https://www.u-tokyo.ac.jp/en/">The University of Tokyo</a>, Laboratory of Biometry and Bioinformatics.</p>
<p><img src="reference/figures/UTlogo-sm.jpg" style="display: block; margin: auto;"></p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Browse source code at <br><a href="https://github.com/ut-biomet/breedSimulatR/">https://​github.com/​ut-biomet/​breedSimulatR/​</a>
</li>
<li>Report a bug at <br><a href="https://github.com/ut-biomet/breedSimulatR/issues">https://​github.com/​ut-biomet/​breedSimulatR/​issues</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Julien Diot <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0002-8738-2034" target="orcid.widget" aria-label="ORCID"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li><a href="authors.html">All authors...</a></li>
</ul>
</div>

  <div class="dev-status">
<h2>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://gitlab.com/juliendiot42/breedSimulatR/commits/master"><img src="https://gitlab.com/juliendiot42/breedSimulatR/badges/master/pipeline.svg" alt="pipeline status"></a></li>
<li><a href="https://codecov.io/gh/ut-biomet/breedSimulatR"><img src="https://codecov.io/gh/ut-biomet/breedSimulatR/branch/master/graph/badge.svg?token=4Uxp1ySLIn" alt="codecov"></a></li>
<li><a href="https://opensource.org/licenses/MIT"><img src="https://img.shields.io/badge/License-MIT-yellow.svg" alt="License: MIT"></a></li>
<li><a href="https://github.com/ut-biomet/breedSimulatR/commits/master"><img src="https://img.shields.io/github/last-commit/ut-biomet/breedSimulatR.svg" alt="last-commit"></a></li>
<li><a href="https://CRAN.R-project.org/package=breedSimulatR"><img src="https://www.r-pkg.org/badges/version/breedSimulatR" alt="CRAN status"></a></li>
<li><a href="https://www.tidyverse.org/lifecycle/#experimental"><img src="https://img.shields.io/badge/lifecycle-experimental-orange.svg" alt="Lifecycle: experimental"></a></li>
<li><a href="https://github.com/ut-biomet/breedSimulatR/actions"><img src="https://github.com/ut-biomet/breedSimulatR/workflows/R-CMD-check/badge.svg" alt="R build status"></a></li>
</ul>
</div>
</div>
</div>


      <footer><div class="copyright">
  <p>Developed by Julien Diot.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
