<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>R6 class representing a phenotypic trait — trait • breedSimulatR</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="R6 class representing a phenotypic trait — trait" />
<meta property="og:description" content="trait object store information about a phenotypic trait" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">breedSimulatR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="../news/index.html">News</a>
</li>
<li>
  <a href="https://github.com/ut-biomet/breedSimulatR/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>R6 class representing a phenotypic trait</h1>
    <small class="dont-index">Source: <a href='https://github.com/ut-biomet/breedSimulatR/blob/master/R/phenotype.R'><code>R/phenotype.R</code></a></small>
    <div class="hidden name"><code>trait.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>trait object store information about a phenotypic trait</p>
    </div>



    <h2 class="hasAnchor" id="public-fields"><a class="anchor" href="#public-fields"></a>Public fields</h2>

    <p><div class="r6-fields"></p><dl>
<dt><code>name</code></dt><dd><p>[string] Name of the trait</p></dd>

<dt><code>class</code></dt><dd><p>"qualitative" or "quantitative" trait ("qualitative" not
implemented yet)</p></dd>

<dt><code>qtn</code></dt><dd><p>[character vector] names of the causal quantitative trait
nucleotides (<code>length(qtn) == length(qtnEff)</code> must be <code>true</code>)</p></dd>

<dt><code>qtnEff</code></dt><dd><p>[numeric vector] quantitative trait nucleotides effects</p></dd>

</dl><p></div></p>
    <h2 class="hasAnchor" id="methods"><a class="anchor" href="#methods"></a>Methods</h2>

    
<h3 class='hasAnchor' id='arguments'><a class='anchor' href='#arguments'></a>Public methods</h3>

<ul>
<li><p><a href='#method-new'><code>trait$new()</code></a></p></li>
<li><p><a href='#method-gv'><code>trait$gv()</code></a></p></li>
<li><p><a href='#method-print'><code>trait$print()</code></a></p></li>
<li><p><a href='#method-clone'><code>trait$clone()</code></a></p></li>
</ul>
<p><hr>
<a id="method-new"></a></p><h3 class='hasAnchor' id='arguments'><a class='anchor' href='#arguments'></a>Method <code>new()</code></h3>
<p>Create a new trait object.</p><h4 class='hasAnchor' id='arguments'><a class='anchor' href='#arguments'></a>Usage</h4>
<p><div class="r"></p><pre><span class='kw'>trait</span><span class='op'>$</span><span class='fu'>new</span>(
  name = <span class='kw'>NULL</span>,
  class = <span class='st'>"quantitative"</span>,
  qtn = <span class='kw'>NULL</span>,
  qtnEff = <span class='kw'>NULL</span>,
  vg = <span class='kw'>NULL</span>,
  afs = <span class='kw'>NULL</span>
)</pre><p></div></p>

<h4 class='hasAnchor' id='arguments'><a class='anchor' href='#arguments'></a>Arguments</h4>
<p><div class="arguments"></p><dl>
<dt><code>name</code></dt><dd><p>[character] name of the trait</p></dd>

<dt><code>class</code></dt><dd><p>"quantitative" or "qualitative"</p></dd>

<dt><code>qtn</code></dt><dd><p>[character vector] list of the quantitative trait
nucleotides names implied in the trait</p></dd>

<dt><code>qtnEff</code></dt><dd><p>[numeric vector] quantitative trait nucleotides effects (see details for more information).</p></dd>

</dl><p></div></p>
<h4 class='hasAnchor' id='arguments'><a class='anchor' href='#arguments'></a>Details</h4>
<p><code>qtn</code> and <code>qtnEff</code> must be in the same order.
  <code>qtnEff[n]</code> must be the QTN effect of the QTN <code>qtn[n]</code>.</p>

<h4 class='hasAnchor' id='arguments'><a class='anchor' href='#arguments'></a>Returns</h4>
<p>A new `trait` object.</p>
<h4 class='hasAnchor' id='arguments'><a class='anchor' href='#arguments'></a>Examples</h4>
<p><div class="r example copy"></p><pre><span class='kw'>mySpec</span> <span class='op'>&lt;-</span> <span class='kw'>specie</span><span class='op'>$</span><span class='fu'>new</span>(nChr = <span class='fl'>10</span>,
                     lchr = <span class='fl'>10</span><span class='op'>^</span><span class='fl'>6</span>,
                     specName = <span class='st'>"Geneticae Exempli"</span>,
                     ploidy = <span class='fl'>2</span>,
                     recombRate = <span class='fl'>3</span><span class='op'>/</span><span class='fl'>10</span><span class='op'>^</span><span class='fl'>6</span>)
<span class='kw'>SNPs</span> <span class='op'>&lt;-</span> <span class='kw'>SNPinfo</span><span class='op'>$</span><span class='fu'>new</span>(SNPcoord = <span class='kw'>exampleData</span><span class='op'>$</span><span class='kw'>snpCoord</span>,
                    specie = <span class='kw'>mySpec</span>)

<span class='kw'>myTrait</span> <span class='op'>&lt;-</span> <span class='kw'>trait</span><span class='op'>$</span><span class='fu'>new</span>(name = <span class='st'>"myTrait"</span>,
                     qtn = <span class='fu'><a href='https://rdrr.io/r/base/sample.html'>sample</a></span>(<span class='kw'>SNPs</span><span class='op'>$</span><span class='kw'>SNPcoord</span><span class='op'>$</span><span class='kw'>SNPid</span>, <span class='fl'>100</span>),
                     qtnEff = <span class='fu'><a href='https://rdrr.io/r/stats/Normal.html'>rnorm</a></span>(<span class='fl'>100</span>, sd = <span class='fl'>0.5</span>))
</pre><p></div></p>

<p><hr>
<a id="method-gv"></a></p><h3 class='hasAnchor' id='arguments'><a class='anchor' href='#arguments'></a>Method <code>gv()</code></h3>
<p>Calculate the genetic values of a population</p><h4 class='hasAnchor' id='arguments'><a class='anchor' href='#arguments'></a>Usage</h4>
<p><div class="r"></p><pre><span class='kw'>trait</span><span class='op'>$</span><span class='fu'>gv</span>(<span class='kw'>pop</span>)</pre><p></div></p>

<h4 class='hasAnchor' id='arguments'><a class='anchor' href='#arguments'></a>Arguments</h4>
<p><div class="arguments"></p><dl>
<dt><code>pop</code></dt><dd><p>[population class] population
(see: <a href='population.html'>population</a>)</p></dd>

</dl><p></div></p>
<h4 class='hasAnchor' id='arguments'><a class='anchor' href='#arguments'></a>Examples</h4>
<p><div class="r example copy"></p><pre><span class='co'># create population</span>
<span class='kw'>example_pop</span> <span class='op'>&lt;-</span> <span class='fu'><a href='createPop.html'>createPop</a></span>(geno = <span class='kw'>exampleData</span><span class='op'>$</span><span class='kw'>genotypes</span>,
                         SNPinfo = <span class='kw'>SNPs</span>,
                         popName = <span class='st'>"Example population"</span>)
<span class='kw'>myTrait</span><span class='op'>$</span><span class='fu'>gv</span>(<span class='kw'>example_pop</span>)
</pre><p></div></p>

<p><hr>
<a id="method-print"></a></p><h3 class='hasAnchor' id='arguments'><a class='anchor' href='#arguments'></a>Method <code><a href='https://rdrr.io/r/base/print.html'>print()</a></code></h3>
<p>Display information about the object</p><h4 class='hasAnchor' id='arguments'><a class='anchor' href='#arguments'></a>Usage</h4>
<p><div class="r"></p><pre><span class='kw'>trait</span><span class='op'>$</span><span class='fu'><a href='https://rdrr.io/r/base/print.html'>print</a></span>()</pre><p></div></p>

<p><hr>
<a id="method-clone"></a></p><h3 class='hasAnchor' id='arguments'><a class='anchor' href='#arguments'></a>Method <code>clone()</code></h3>
<p>The objects of this class are cloneable with this method.</p><h4 class='hasAnchor' id='arguments'><a class='anchor' href='#arguments'></a>Usage</h4>
<p><div class="r"></p><pre><span class='kw'>trait</span><span class='op'>$</span><span class='fu'>clone</span>(deep = <span class='fl'>FALSE</span>)</pre><p></div></p>

<h4 class='hasAnchor' id='arguments'><a class='anchor' href='#arguments'></a>Arguments</h4>
<p><div class="arguments"></p><dl>
<dt><code>deep</code></dt><dd><p>Whether to make a deep clone.</p></dd>

</dl><p></div></p>



    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'>
<span class='co'>## ------------------------------------------------</span>
<span class='co'>## Method `trait$new`</span>
<span class='co'>## ------------------------------------------------</span>

<span class='kw'>mySpec</span> <span class='op'>&lt;-</span> <span class='kw'>specie</span><span class='op'>$</span><span class='fu'>new</span>(nChr = <span class='fl'>10</span>,
                     lchr = <span class='fl'>10</span><span class='op'>^</span><span class='fl'>6</span>,
                     specName = <span class='st'>"Geneticae Exempli"</span>,
                     ploidy = <span class='fl'>2</span>,
                     recombRate = <span class='fl'>3</span><span class='op'>/</span><span class='fl'>10</span><span class='op'>^</span><span class='fl'>6</span>)
</div><div class='output co'>#&gt; A new species has emerged: Geneticae Exempli !
#&gt; </div><div class='input'><span class='kw'>SNPs</span> <span class='op'>&lt;-</span> <span class='kw'>SNPinfo</span><span class='op'>$</span><span class='fu'>new</span>(SNPcoord = <span class='kw'>exampleData</span><span class='op'>$</span><span class='kw'>snpCoord</span>,
                    specie = <span class='kw'>mySpec</span>)

<span class='kw'>myTrait</span> <span class='op'>&lt;-</span> <span class='kw'>trait</span><span class='op'>$</span><span class='fu'>new</span>(name = <span class='st'>"myTrait"</span>,
                     qtn = <span class='fu'><a href='https://rdrr.io/r/base/sample.html'>sample</a></span>(<span class='kw'>SNPs</span><span class='op'>$</span><span class='kw'>SNPcoord</span><span class='op'>$</span><span class='kw'>SNPid</span>, <span class='fl'>100</span>),
                     qtnEff = <span class='fu'><a href='https://rdrr.io/r/stats/Normal.html'>rnorm</a></span>(<span class='fl'>100</span>, sd = <span class='fl'>0.5</span>))

<span class='co'>## ------------------------------------------------</span>
<span class='co'>## Method `trait$gv`</span>
<span class='co'>## ------------------------------------------------</span>

<span class='co'># create population</span>
<span class='kw'>example_pop</span> <span class='op'>&lt;-</span> <span class='fu'><a href='createPop.html'>createPop</a></span>(geno = <span class='kw'>exampleData</span><span class='op'>$</span><span class='kw'>genotypes</span>,
                         SNPinfo = <span class='kw'>SNPs</span>,
                         popName = <span class='st'>"Example population"</span>)
</div><div class='output co'>#&gt; Create population: Initialisation...
#&gt; Create population: Create individuals...
#&gt; 1%2%3%4%5%6%7%8%9%10%11%12%13%14%15%16%17%18%19%20%21%22%23%24%25%26%27%28%29%30%31%32%33%34%35%36%37%38%39%40%41%42%43%44%45%46%47%48%49%50%51%52%53%54%55%56%57%58%59%60%61%62%63%64%65%66%67%68%69%70%71%72%73%74%75%76%77%78%79%80%81%82%83%84%85%86%87%88%89%90%91%92%93%94%95%96%97%98%99%100%
#&gt; Create population: Create population object...
#&gt; Create population: Add individuals...
#&gt; 1%2%3%4%5%6%7%8%9%10%11%12%13%14%15%16%17%18%19%20%21%22%23%24%25%26%27%28%29%30%31%32%33%34%35%36%37%38%39%40%41%42%43%44%45%46%47%48%49%50%51%52%53%54%55%56%57%58%59%60%61%62%63%64%65%66%67%68%69%70%71%72%73%74%75%76%77%78%79%80%81%82%83%84%85%86%87%88%89%90%91%92%93%94%95%96%97%98%99%100%
#&gt; A new population created: Example population !</div><div class='input'><span class='kw'>myTrait</span><span class='op'>$</span><span class='fu'>gv</span>(<span class='kw'>example_pop</span>)
</div><div class='output co'>#&gt;                 [,1]
#&gt; Coll0001   2.7178022
#&gt; Coll0002  -7.3329499
#&gt; Coll0003  -4.1501852
#&gt; Coll0004  -8.3372600
#&gt; Coll0005  -1.8441642
#&gt; Coll0006  -4.0918945
#&gt; Coll0007  -3.6628397
#&gt; Coll0008  -4.4268337
#&gt; Coll0009  -5.5224824
#&gt; Coll0010 -11.7017875
#&gt; Coll0011  -8.8740015
#&gt; Coll0012  -3.6034233
#&gt; Coll0013   1.6110625
#&gt; Coll0014  -3.3700490
#&gt; Coll0015 -14.5840378
#&gt; Coll0016  -5.3113940
#&gt; Coll0017   1.6212251
#&gt; Coll0018  -4.9999024
#&gt; Coll0019   0.8756595
#&gt; Coll0020   0.7398596
#&gt; Coll0021   3.3453433
#&gt; Coll0022  -4.7799026
#&gt; Coll0023 -10.3881525
#&gt; Coll0024   0.4525894
#&gt; Coll0025  -5.5052386
#&gt; Coll0026  -2.1048737
#&gt; Coll0027   3.6202163
#&gt; Coll0028  -5.2853341
#&gt; Coll0029   1.9311643
#&gt; Coll0030  -1.4158721
#&gt; Coll0031  -1.2596829
#&gt; Coll0032   3.6101593
#&gt; Coll0033   1.3874153
#&gt; Coll0034  -3.8616874
#&gt; Coll0035  -3.2654685
#&gt; Coll0036 -12.2465223
#&gt; Coll0037  -6.1403798
#&gt; Coll0038  -0.1519302
#&gt; Coll0039  -6.3444200
#&gt; Coll0040  -1.8021201
#&gt; Coll0041 -17.3238240
#&gt; Coll0042  -5.3367852
#&gt; Coll0043  -2.5954757
#&gt; Coll0044  -4.5675512
#&gt; Coll0045  -9.9258339
#&gt; Coll0046 -12.0442862
#&gt; Coll0047   1.1812372
#&gt; Coll0048   6.5132270
#&gt; Coll0049  -4.7611605
#&gt; Coll0050  -7.1648308
#&gt; Coll0051  -7.9979857
#&gt; Coll0052  -4.9671098
#&gt; Coll0053  -0.4858822
#&gt; Coll0054   0.4836165
#&gt; Coll0055  -4.4505183
#&gt; Coll0056  -9.6203308
#&gt; Coll0057  -3.8084546
#&gt; Coll0058 -11.4741095
#&gt; Coll0059  -9.8558876
#&gt; Coll0060  -2.9118392
#&gt; Coll0061   2.7959231
#&gt; Coll0062  -0.2015518
#&gt; Coll0063  -5.3976787
#&gt; Coll0064  -7.2025390
#&gt; Coll0065   0.4722737
#&gt; Coll0066   1.6360332
#&gt; Coll0067   6.7475464
#&gt; Coll0068   1.9196162
#&gt; Coll0069 -10.7337546
#&gt; Coll0070  -5.2331585
#&gt; Coll0071  -5.4624479
#&gt; Coll0072   5.4495686
#&gt; Coll0073  -1.8725235
#&gt; Coll0074  -3.1665783
#&gt; Coll0075  -7.0840061
#&gt; Coll0076  -1.2823682
#&gt; Coll0077  -5.4422462
#&gt; Coll0078  -5.8541814
#&gt; Coll0079  -9.1301785
#&gt; Coll0080 -11.8489319
#&gt; Coll0081  -0.9426233
#&gt; Coll0082  -4.7687268
#&gt; Coll0083   2.6007378
#&gt; Coll0084  -6.4760691
#&gt; Coll0085  -1.2346903
#&gt; Coll0086  -8.0501361
#&gt; Coll0087  -2.5263611
#&gt; Coll0088  -2.8387750
#&gt; Coll0089  -3.1069742
#&gt; Coll0090  -2.5833045
#&gt; Coll0091   2.4134636
#&gt; Coll0092  -4.3248956
#&gt; Coll0093  -0.4692350
#&gt; Coll0094  -3.0899308
#&gt; Coll0095  -8.0044591
#&gt; Coll0096  -9.4520829
#&gt; Coll0097  -0.1481705
#&gt; Coll0098  -5.8151304
#&gt; Coll0099  -9.2678015
#&gt; Coll0100  -9.8788327</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Julien Diot.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.9000.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


