<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>R6 class representing a phenotypic trait — trait • breedSimulatR</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="R6 class representing a phenotypic trait — trait" />
<meta property="og:description" content="trait object store information about a phenotypic trait" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">breedSimulatR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.2.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="../news/index.html">News</a>
</li>
<li>
  <a href="https://github.com/ut-biomet/breedSimulatR/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>R6 class representing a phenotypic trait</h1>
    <small class="dont-index">Source: <a href='https://github.com/ut-biomet/breedSimulatR/blob/master/R/phenotype.R'><code>R/phenotype.R</code></a></small>
    <div class="hidden name"><code>trait.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>trait object store information about a phenotypic trait</p>
    </div>



    <h2 class="hasAnchor" id="public-fields"><a class="anchor" href="#public-fields"></a>Public fields</h2>

    <p><div class="r6-fields"></p><dl>
<dt><code>name</code></dt><dd><p>[string] Name of the trait</p></dd>

<dt><code>class</code></dt><dd><p>"qualitative" or "quantitative" trait ("qualitative" not
implemented yet)</p></dd>

<dt><code>qtn</code></dt><dd><p>[character vector] names of the causal quantitative trait
nucleotides (<code>length(qtn) == length(qtnEff)</code> must be <code>true</code>)</p></dd>

<dt><code>qtnEff</code></dt><dd><p>[numeric vector] quantitative trait nucleotides effects</p></dd>

</dl><p></div></p>
    <h2 class="hasAnchor" id="methods"><a class="anchor" href="#methods"></a>Methods</h2>

    
<h3 class='hasAnchor' id='arguments'><a class='anchor' href='#arguments'></a>Public methods</h3>

<ul>
<li><p><a href='#method-new'><code>trait$new()</code></a></p></li>
<li><p><a href='#method-gv'><code>trait$gv()</code></a></p></li>
<li><p><a href='#method-print'><code>trait$print()</code></a></p></li>
<li><p><a href='#method-clone'><code>trait$clone()</code></a></p></li>
</ul>
<p><hr>
<a id="method-new"></a></p><h3 class='hasAnchor' id='arguments'><a class='anchor' href='#arguments'></a>Method <code>new()</code></h3>
<p>Create a new trait object.</p><h4 class='hasAnchor' id='arguments'><a class='anchor' href='#arguments'></a>Usage</h4>
<p><div class="r"></p><pre><span class='va'>trait</span><span class='op'>$</span><span class='fu'>new</span><span class='op'>(</span>
  name <span class='op'>=</span> <span class='cn'>NULL</span>,
  class <span class='op'>=</span> <span class='st'>"quantitative"</span>,
  qtn <span class='op'>=</span> <span class='cn'>NULL</span>,
  qtnEff <span class='op'>=</span> <span class='cn'>NULL</span>,
  vg <span class='op'>=</span> <span class='cn'>NULL</span>,
  afs <span class='op'>=</span> <span class='cn'>NULL</span>
<span class='op'>)</span></pre><p></div></p>

<h4 class='hasAnchor' id='arguments'><a class='anchor' href='#arguments'></a>Arguments</h4>
<p><div class="arguments"></p><dl>
<dt><code>name</code></dt><dd><p>[character] name of the trait</p></dd>

<dt><code>class</code></dt><dd><p>"quantitative" or "qualitative"</p></dd>

<dt><code>qtn</code></dt><dd><p>[character vector] list of the quantitative trait
nucleotides names implied in the trait</p></dd>

<dt><code>qtnEff</code></dt><dd><p>[numeric vector] quantitative trait nucleotides effects
(see details for more information).</p></dd>

</dl><p></div></p>
<h4 class='hasAnchor' id='arguments'><a class='anchor' href='#arguments'></a>Details</h4>
<p><code>qtn</code> and <code>qtnEff</code> must be in the same order.
  <code>qtnEff[n]</code> must be the QTN effect of the QTN <code>qtn[n]</code>.</p>

<h4 class='hasAnchor' id='arguments'><a class='anchor' href='#arguments'></a>Returns</h4>
<p>A new `trait` object.</p>
<h4 class='hasAnchor' id='arguments'><a class='anchor' href='#arguments'></a>Examples</h4>
<p><div class="r example copy"></p><pre><span class='va'>mySpec</span> <span class='op'>&lt;-</span> <span class='va'><a href='specie.html'>specie</a></span><span class='op'>$</span><span class='fu'>new</span><span class='op'>(</span>nChr <span class='op'>=</span> <span class='fl'>10</span>,
                     lchr <span class='op'>=</span> <span class='fl'>10</span><span class='op'>^</span><span class='fl'>6</span>,
                     lchrCm <span class='op'>=</span> <span class='fl'>100</span>,
                     specName <span class='op'>=</span> <span class='st'>"Geneticae Exempli"</span>,
                     ploidy <span class='op'>=</span> <span class='fl'>2</span><span class='op'>)</span>
<span class='va'>SNPs</span> <span class='op'>&lt;-</span> <span class='va'><a href='SNPinfo.html'>SNPinfo</a></span><span class='op'>$</span><span class='fu'>new</span><span class='op'>(</span>SNPcoord <span class='op'>=</span> <span class='va'>exampleData</span><span class='op'>$</span><span class='va'>snpCoord</span>,
                    specie <span class='op'>=</span> <span class='va'>mySpec</span><span class='op'>)</span>

<span class='va'>myTrait</span> <span class='op'>&lt;-</span> <span class='va'>trait</span><span class='op'>$</span><span class='fu'>new</span><span class='op'>(</span>name <span class='op'>=</span> <span class='st'>"myTrait"</span>,
                     qtn <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/sample.html'>sample</a></span><span class='op'>(</span><span class='va'>SNPs</span><span class='op'>$</span><span class='va'>SNPcoord</span><span class='op'>$</span><span class='va'>SNPid</span>, <span class='fl'>100</span><span class='op'>)</span>,
                     qtnEff <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/stats/Normal.html'>rnorm</a></span><span class='op'>(</span><span class='fl'>100</span>, sd <span class='op'>=</span> <span class='fl'>0.5</span><span class='op'>)</span><span class='op'>)</span>
</pre><p></div></p>

<p><hr>
<a id="method-gv"></a></p><h3 class='hasAnchor' id='arguments'><a class='anchor' href='#arguments'></a>Method <code>gv()</code></h3>
<p>Calculate the genetic values of a population</p><h4 class='hasAnchor' id='arguments'><a class='anchor' href='#arguments'></a>Usage</h4>
<p><div class="r"></p><pre><span class='va'>trait</span><span class='op'>$</span><span class='fu'>gv</span><span class='op'>(</span><span class='va'>pop</span><span class='op'>)</span></pre><p></div></p>

<h4 class='hasAnchor' id='arguments'><a class='anchor' href='#arguments'></a>Arguments</h4>
<p><div class="arguments"></p><dl>
<dt><code>pop</code></dt><dd><p>[population class] population
(see: <a href='population.html'>population</a>)</p></dd>

</dl><p></div></p>
<h4 class='hasAnchor' id='arguments'><a class='anchor' href='#arguments'></a>Examples</h4>
<p><div class="r example copy"></p><pre><span class='co'># create population</span>
<span class='va'>example_pop</span> <span class='op'>&lt;-</span> <span class='fu'><a href='createPop.html'>createPop</a></span><span class='op'>(</span>geno <span class='op'>=</span> <span class='va'>exampleData</span><span class='op'>$</span><span class='va'>genotypes</span>,
                         SNPinfo <span class='op'>=</span> <span class='va'>SNPs</span>,
                         popName <span class='op'>=</span> <span class='st'>"Example population"</span><span class='op'>)</span>
<span class='va'>myTrait</span><span class='op'>$</span><span class='fu'>gv</span><span class='op'>(</span><span class='va'>example_pop</span><span class='op'>)</span>
</pre><p></div></p>

<p><hr>
<a id="method-print"></a></p><h3 class='hasAnchor' id='arguments'><a class='anchor' href='#arguments'></a>Method <code><a href='https://rdrr.io/r/base/print.html'>print()</a></code></h3>
<p>Display information about the object</p><h4 class='hasAnchor' id='arguments'><a class='anchor' href='#arguments'></a>Usage</h4>
<p><div class="r"></p><pre><span class='va'>trait</span><span class='op'>$</span><span class='fu'>print</span><span class='op'>(</span><span class='op'>)</span></pre><p></div></p>

<p><hr>
<a id="method-clone"></a></p><h3 class='hasAnchor' id='arguments'><a class='anchor' href='#arguments'></a>Method <code>clone()</code></h3>
<p>The objects of this class are cloneable with this method.</p><h4 class='hasAnchor' id='arguments'><a class='anchor' href='#arguments'></a>Usage</h4>
<p><div class="r"></p><pre><span class='va'>trait</span><span class='op'>$</span><span class='fu'>clone</span><span class='op'>(</span>deep <span class='op'>=</span> <span class='cn'>FALSE</span><span class='op'>)</span></pre><p></div></p>

<h4 class='hasAnchor' id='arguments'><a class='anchor' href='#arguments'></a>Arguments</h4>
<p><div class="arguments"></p><dl>
<dt><code>deep</code></dt><dd><p>Whether to make a deep clone.</p></dd>

</dl><p></div></p>



    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'>
<span class='co'>## ------------------------------------------------</span>
<span class='co'>## Method `trait$new`</span>
<span class='co'>## ------------------------------------------------</span>

<span class='va'>mySpec</span> <span class='op'>&lt;-</span> <span class='va'><a href='specie.html'>specie</a></span><span class='op'>$</span><span class='fu'>new</span><span class='op'>(</span>nChr <span class='op'>=</span> <span class='fl'>10</span>,
                     lchr <span class='op'>=</span> <span class='fl'>10</span><span class='op'>^</span><span class='fl'>6</span>,
                     lchrCm <span class='op'>=</span> <span class='fl'>100</span>,
                     specName <span class='op'>=</span> <span class='st'>"Geneticae Exempli"</span>,
                     ploidy <span class='op'>=</span> <span class='fl'>2</span><span class='op'>)</span>
</div><div class='output co'>#&gt; A new species has emerged: Geneticae Exempli !
#&gt; </div><div class='input'><span class='va'>SNPs</span> <span class='op'>&lt;-</span> <span class='va'><a href='SNPinfo.html'>SNPinfo</a></span><span class='op'>$</span><span class='fu'>new</span><span class='op'>(</span>SNPcoord <span class='op'>=</span> <span class='va'>exampleData</span><span class='op'>$</span><span class='va'>snpCoord</span>,
                    specie <span class='op'>=</span> <span class='va'>mySpec</span><span class='op'>)</span>

<span class='va'>myTrait</span> <span class='op'>&lt;-</span> <span class='va'>trait</span><span class='op'>$</span><span class='fu'>new</span><span class='op'>(</span>name <span class='op'>=</span> <span class='st'>"myTrait"</span>,
                     qtn <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/sample.html'>sample</a></span><span class='op'>(</span><span class='va'>SNPs</span><span class='op'>$</span><span class='va'>SNPcoord</span><span class='op'>$</span><span class='va'>SNPid</span>, <span class='fl'>100</span><span class='op'>)</span>,
                     qtnEff <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/stats/Normal.html'>rnorm</a></span><span class='op'>(</span><span class='fl'>100</span>, sd <span class='op'>=</span> <span class='fl'>0.5</span><span class='op'>)</span><span class='op'>)</span>

<span class='co'>## ------------------------------------------------</span>
<span class='co'>## Method `trait$gv`</span>
<span class='co'>## ------------------------------------------------</span>

<span class='co'># create population</span>
<span class='va'>example_pop</span> <span class='op'>&lt;-</span> <span class='fu'><a href='createPop.html'>createPop</a></span><span class='op'>(</span>geno <span class='op'>=</span> <span class='va'>exampleData</span><span class='op'>$</span><span class='va'>genotypes</span>,
                         SNPinfo <span class='op'>=</span> <span class='va'>SNPs</span>,
                         popName <span class='op'>=</span> <span class='st'>"Example population"</span><span class='op'>)</span>
</div><div class='output co'>#&gt; Create population: Initialisation...
#&gt; Create population: Create individuals...
#&gt; 1%2%3%4%5%6%7%8%9%10%11%12%13%14%15%16%17%18%19%20%21%22%23%24%25%26%27%28%29%30%31%32%33%34%35%36%37%38%39%40%41%42%43%44%45%46%47%48%49%50%51%52%53%54%55%56%57%58%59%60%61%62%63%64%65%66%67%68%69%70%71%72%73%74%75%76%77%78%79%80%81%82%83%84%85%86%87%88%89%90%91%92%93%94%95%96%97%98%99%100%
#&gt; Create population: Create population object...
#&gt; Create population: Add individuals...
#&gt; 1%2%3%4%5%6%7%8%9%10%11%12%13%14%15%16%17%18%19%20%21%22%23%24%25%26%27%28%29%30%31%32%33%34%35%36%37%38%39%40%41%42%43%44%45%46%47%48%49%50%51%52%53%54%55%56%57%58%59%60%61%62%63%64%65%66%67%68%69%70%71%72%73%74%75%76%77%78%79%80%81%82%83%84%85%86%87%88%89%90%91%92%93%94%95%96%97%98%99%100%
#&gt; A new population created: Example population !</div><div class='input'><span class='va'>myTrait</span><span class='op'>$</span><span class='fu'>gv</span><span class='op'>(</span><span class='va'>example_pop</span><span class='op'>)</span>
</div><div class='output co'>#&gt;                [,1]
#&gt; Coll0001 -4.2158834
#&gt; Coll0002  8.3548092
#&gt; Coll0003  3.7723037
#&gt; Coll0004  3.8181822
#&gt; Coll0005 12.3965430
#&gt; Coll0006 -4.4802869
#&gt; Coll0007  2.3991851
#&gt; Coll0008 12.0305083
#&gt; Coll0009  1.0671912
#&gt; Coll0010 -2.8938691
#&gt; Coll0011 -0.2320498
#&gt; Coll0012  1.4804435
#&gt; Coll0013 -3.9327907
#&gt; Coll0014 -1.1846558
#&gt; Coll0015 16.2560498
#&gt; Coll0016  3.1415145
#&gt; Coll0017 -0.1058846
#&gt; Coll0018  3.4286479
#&gt; Coll0019  4.3406178
#&gt; Coll0020  6.7881144
#&gt; Coll0021  6.5257012
#&gt; Coll0022  7.8131965
#&gt; Coll0023  6.3831769
#&gt; Coll0024  4.0374779
#&gt; Coll0025  8.3855127
#&gt; Coll0026  5.2369238
#&gt; Coll0027 -6.3388320
#&gt; Coll0028 -2.9164099
#&gt; Coll0029  2.1408798
#&gt; Coll0030 -2.2292685
#&gt; Coll0031  5.8138775
#&gt; Coll0032  4.9860527
#&gt; Coll0033 -1.9387855
#&gt; Coll0034  1.4150423
#&gt; Coll0035  2.0259025
#&gt; Coll0036  4.0077140
#&gt; Coll0037  8.0265737
#&gt; Coll0038 -0.3183030
#&gt; Coll0039 12.7400834
#&gt; Coll0040  8.3389574
#&gt; Coll0041 17.8613660
#&gt; Coll0042  0.7842534
#&gt; Coll0043 12.7257600
#&gt; Coll0044 10.5740236
#&gt; Coll0045  8.4581743
#&gt; Coll0046  9.4841849
#&gt; Coll0047  0.6052785
#&gt; Coll0048 10.7164510
#&gt; Coll0049  4.4981076
#&gt; Coll0050  8.9119367
#&gt; Coll0051  1.3075987
#&gt; Coll0052 -1.8619359
#&gt; Coll0053  4.3766028
#&gt; Coll0054  2.7569067
#&gt; Coll0055  2.8419351
#&gt; Coll0056 -7.4289293
#&gt; Coll0057 10.2420323
#&gt; Coll0058  3.0596717
#&gt; Coll0059  3.0606273
#&gt; Coll0060  8.5380842
#&gt; Coll0061 -6.1442407
#&gt; Coll0062 -6.4541820
#&gt; Coll0063  1.8867622
#&gt; Coll0064  3.5846764
#&gt; Coll0065 -3.1316436
#&gt; Coll0066  1.4609023
#&gt; Coll0067  2.0365347
#&gt; Coll0068  9.1110817
#&gt; Coll0069  3.6662765
#&gt; Coll0070  8.5527853
#&gt; Coll0071 -2.3053222
#&gt; Coll0072  5.9575512
#&gt; Coll0073  1.0461343
#&gt; Coll0074  7.8330840
#&gt; Coll0075  1.8686989
#&gt; Coll0076  2.7073347
#&gt; Coll0077  7.5065292
#&gt; Coll0078  6.8937198
#&gt; Coll0079  4.8387148
#&gt; Coll0080  7.3985669
#&gt; Coll0081  8.1122827
#&gt; Coll0082  0.3961473
#&gt; Coll0083 15.3687560
#&gt; Coll0084  1.0182599
#&gt; Coll0085  9.3036826
#&gt; Coll0086  0.9711091
#&gt; Coll0087 10.5160808
#&gt; Coll0088  7.9832562
#&gt; Coll0089  8.5707037
#&gt; Coll0090 -2.0025086
#&gt; Coll0091  9.4172620
#&gt; Coll0092  3.5657786
#&gt; Coll0093  6.3781416
#&gt; Coll0094 -6.6802091
#&gt; Coll0095  9.2803725
#&gt; Coll0096  3.8545229
#&gt; Coll0097  5.8371670
#&gt; Coll0098  7.7040458
#&gt; Coll0099  7.3843881
#&gt; Coll0100  6.3786784</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Julien Diot.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.9000.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


