<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-US">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>`breedSimulatR` Example • breedSimulatR</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="`breedSimulatR` Example">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">breedSimulatR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.3.2</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../index.html"><span class="fa fas fa-home fa-lg"></span></a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/example.html">`breedSimulatR` Example</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><a class="nav-link" href="../news/index.html">News</a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/ut-biomet/breedSimulatR/"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>`breedSimulatR` Example</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/ut-biomet/breedSimulatR/blob/master/vignettes/example.Rmd" class="external-link"><code>vignettes/example.Rmd</code></a></small>
      <div class="d-none name"><code>example.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p><code>breedSimulatR</code> is R package providing classes and
functions to simulate breeding schemes. It have been design to be
intuitive and easy to use.</p>
<p>The Object-Oriented Programming (OOP) system used in this package is
<a href="https://CRAN.R-project.org/package=R6" class="external-link">R6</a>. Information
about R6 OOP system can be found at <a href="https://adv-r.hadley.nz/r6.html" class="external-link">chapter 14 of Hadley Wickham’s
book <em>Advanced R</em></a>.</p>
<p>This vignette shows how to use the <code>breedSimulatR</code> package
thought basic examples.</p>
<div class="section level3">
<h3 id="breedsimulatrs-classes">
<code>breedSimulatR</code>’s classes<a class="anchor" aria-label="anchor" href="#breedsimulatrs-classes"></a>
</h3>
<p>First let’s have a look on the main classes provided by
<code>breedSimulatR</code>:</p>
<ul>
<li>
<a href="https://ut-biomet.github.io/breedSimulatR/reference/specie.html"><code>Specie</code></a>:
which store specific information about the studied specie like number of
chromosomes, length of chromosomes in base pairs and centimorgans, their
names</li>
<li>
<a href="https://ut-biomet.github.io/breedSimulatR/reference/SNPinfo.html"><code>SNPinfo</code></a>:
which store specific information about a set of SNP markers like
physical positions, linkage map position, IDs</li>
<li>
<a href="https://ut-biomet.github.io/breedSimulatR/reference/individual.html"><code>individuals</code></a>:
which store specific information about an individual like its parents,
its haplotype. This class have also a method to generate gametes.</li>
<li>
<a href="https://ut-biomet.github.io/breedSimulatR/reference/population.html"><code>population</code></a>:
which store specific information several individuals like their
genotypes, allele frequency</li>
<li>
<a href="https://ut-biomet.github.io/breedSimulatR/reference/trait.html"><code>trait</code></a>:
which store specific information about a trait like the QTN involved,
the QTN’s effects</li>
<li>
<a href="https://ut-biomet.github.io/breedSimulatR/reference/phenotyper.html"><code>phenotyper</code></a>:
This object is used to simulate phenotyping experiments and store store
some specific information like the variance of the environmental effect
and have method to phenotype individuals. Several phenotyper can be
created to represents several environments.</li>
</ul>
<p>A exhaustive list can be fond on the package’s <a href="https://ut-biomet.github.io/breedSimulatR/reference/index.html">web
site</a>.</p>
</div>
<div class="section level3">
<h3 id="example-data">Example data<a class="anchor" aria-label="anchor" href="#example-data"></a>
</h3>
<p>The package contains some example data that we will use for these
examples. These data are stored in the variable <code>exampleData</code>
of the package.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/ut-biomet/breedSimulatR" class="external-link">breedSimulatR</a></span><span class="op">)</span></span>
<span><span class="va">exampleData</span></span></code></pre></div>
<p><code>exampleData</code> is a list containing 3 elements:</p>
<ul>
<li>
<code>exampleData$genotypes</code>: contains some genotypic.</li>
<li>
<code>exampleData$snpCoord</code>: contains the coordinates of the
SNP markers.</li>
<li>
<code>exampleData$snpEffects</code>: contains the “true” effects of
the SNP markers for a quantitative trait based on an additive
architecture.</li>
</ul>
<p>It represents information about <code>.nInd</code> individuals and
<code>.nSNP</code> markers on <code>.nChr</code>.</p>
<p>These data are fictitious and come from the serious game <a href="https://github.com/timflutre/PlantBreedGame" class="external-link">“PlantBreedGame”</a>
<span class="citation">(Flutre, Diot, and David 2019)</span>.</p>
</div>
</div>
<div class="section level2">
<h2 id="simulation-initialization">Simulation Initialization<a class="anchor" aria-label="anchor" href="#simulation-initialization"></a>
</h2>
<p>In order to run some simulation, we must first initialize some
objects: a <code>specie</code>, SNP information (<code>SNPinfo</code>),
a <code>population</code>, some <code>traits</code> and a
<code>phenotyper</code>.</p>
<div class="section level3">
<h3 id="specie-specification">Specie specification<a class="anchor" aria-label="anchor" href="#specie-specification"></a>
</h3>
<p>The example’s specie, let’s name it <em>Statisticae exempli</em>,
have <strong>10 chromosomes</strong> with the same length of around
<strong>1e+06</strong> base pairs for <strong>100</strong>
centimorgans.</p>
<p>The function <code>specie$new</code> will create a new
<code>specie</code> object <code>specie_statEx</code>.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># create specie object</span></span>
<span><span class="va">specie_statEx</span> <span class="op">&lt;-</span> <span class="va"><a href="../reference/specie.html">specie</a></span><span class="op">$</span><span class="fu">new</span><span class="op">(</span></span>
<span>  specName <span class="op">=</span> <span class="st">"Statisticae exempli"</span>,</span>
<span>  nChr <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  lchr <span class="op">=</span> <span class="fl">1e6</span>,</span>
<span>  lchrCm <span class="op">=</span> <span class="fl">100</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; A new species has emerged: Statisticae exempli !</span></span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">specie_statEx</span><span class="op">)</span></span>
<span><span class="co">#&gt; Name: Statisticae exempli</span></span>
<span><span class="co">#&gt; Number of Chromosomes: 10</span></span>
<span><span class="co">#&gt; Ploidy: 2</span></span>
<span><span class="co">#&gt; Chromosome length:</span></span>
<span><span class="co">#&gt;       chrNames chrLength chrLengthCm</span></span>
<span><span class="co">#&gt; Chr01    Chr01     1e+06         100</span></span>
<span><span class="co">#&gt; Chr02    Chr02     1e+06         100</span></span>
<span><span class="co">#&gt; Chr03    Chr03     1e+06         100</span></span>
<span><span class="co">#&gt; Chr04    Chr04     1e+06         100</span></span>
<span><span class="co">#&gt; Chr05    Chr05     1e+06         100</span></span>
<span><span class="co">#&gt; Chr06    Chr06     1e+06         100</span></span>
<span><span class="co">#&gt; Chr07    Chr07     1e+06         100</span></span>
<span><span class="co">#&gt; Chr08    Chr08     1e+06         100</span></span>
<span><span class="co">#&gt;  [ reached 'max' / getOption("max.print") -- omitted 2 rows ]</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="snp-specification">SNP specification<a class="anchor" aria-label="anchor" href="#snp-specification"></a>
</h3>
<p>In order to simulate the crossing between individuals, we need
information about the positions of the genotypic markers used in the
simulation. We will create for that the object <code>SNPinfo</code>
using the function <code>SNPinfo$new</code>.</p>
<ul>
<li>
<code>SNPcoord</code>: a <code>data.frame</code> with one line per
marker and 4 columns:
<ul>
<li>
<code>chr</code>: Chromosome holding the SNP</li>
<li>
<code>physPos</code>: SNP physical position on the chromosome</li>
<li>
<code>linkMapPos</code>: SNP linkage map position on the
chromosome</li>
<li>
<code>SNPid</code>: SNP’s IDs</li>
</ul>
</li>
<li>
<code>specie</code>: an object of class specie.</li>
</ul>
<p>We can create the <code>SNPinfo</code> object using
<code>exampleData$snpCoord</code> which is a <code>data.frame</code>
matching the requirement of <code>SNPinfo$new</code>:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># data preview</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">exampleData</span><span class="op">$</span><span class="va">snpCoord</span><span class="op">)</span></span>
<span><span class="co">#&gt;      chr physPos linkMapPos    SNPid</span></span>
<span><span class="co">#&gt; 2  Chr01  937638   86.81546 snp03760</span></span>
<span><span class="co">#&gt; 6  Chr01  654763   56.51842 snp02674</span></span>
<span><span class="co">#&gt; 10 Chr01  181658   30.62157 snp00721</span></span>
<span><span class="co">#&gt; 18 Chr01  230126   35.47120 snp00948</span></span>
<span><span class="co">#&gt; 26 Chr01  420637   46.98455 snp01620</span></span>
<span><span class="co">#&gt; 29 Chr01  467620   48.80905 snp01790</span></span></code></pre></div>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># create SNPinfo object</span></span>
<span><span class="va">SNPs</span> <span class="op">&lt;-</span> <span class="va"><a href="../reference/SNPinfo.html">SNPinfo</a></span><span class="op">$</span><span class="fu">new</span><span class="op">(</span></span>
<span>  SNPcoord <span class="op">=</span> <span class="va">exampleData</span><span class="op">$</span><span class="va">snpCoord</span>,</span>
<span>  specie <span class="op">=</span> <span class="va">specie_statEx</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">SNPs</span><span class="op">)</span></span>
<span><span class="co">#&gt; specie: Statisticae exempli</span></span>
<span><span class="co">#&gt; 3333 Markers on 10 chromosomes :</span></span>
<span><span class="co">#&gt; Chr01 Chr02 Chr03 Chr04 Chr05 Chr06 Chr07 Chr08 Chr09 Chr10 </span></span>
<span><span class="co">#&gt;   415   247   425   322   381   269   238   355   342   339 </span></span>
<span><span class="co">#&gt; SNPcoord:</span></span>
<span><span class="co">#&gt;            chr    SNPid physPos linkMapPos</span></span>
<span><span class="co">#&gt; snp00006 Chr01 snp00006    2068  0.4911687</span></span>
<span><span class="co">#&gt; snp00009 Chr01 snp00009    2708  0.6423784</span></span>
<span><span class="co">#&gt; snp00011 Chr01 snp00011    2782  0.6598378</span></span>
<span><span class="co">#&gt; snp00018 Chr01 snp00018    4159  0.9838113</span></span>
<span><span class="co">#&gt; snp00026 Chr01 snp00026    6917  1.6275084</span></span>
<span><span class="co">#&gt; snp00031 Chr01 snp00031    7814  1.8353769</span></span>
<span><span class="co">#&gt;  [ reached 'max' / getOption("max.print") -- omitted 3327 rows ]</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="population-initialization">Population initialization<a class="anchor" aria-label="anchor" href="#population-initialization"></a>
</h3>
<p>We can now generate an initial population from genotypic data. We
will use the function <code>createPop</code>. This function will create
a <code>population</code> object from at least two arguments:</p>
<ul>
<li>
<code>geno</code>: a <code>data.frame</code> of the genotypic data
of homozygotes individuals</li>
<li>
<code>SNPinfo</code>: a <code>SNPinfo</code> object containing the
information related to the markers of the genotypic data</li>
</ul>
<p>We can create the initial <code>population</code> object using
<code>exampleData$genotypes</code> which is a <code>data.frame</code>
matching the requirement of <code>geno</code>:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># data preview</span></span>
<span><span class="va">exampleData</span><span class="op">$</span><span class="va">genotypes</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span>
<span><span class="co">#&gt;          snp00006 snp00009 snp00011 snp00018 snp00026</span></span>
<span><span class="co">#&gt; Coll0001        2        2        2        0        2</span></span>
<span><span class="co">#&gt; Coll0002        0        2        2        2        0</span></span>
<span><span class="co">#&gt; Coll0003        2        2        2        0        2</span></span></code></pre></div>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># create population object</span></span>
<span><span class="va">initPop</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createPop.html">createPop</a></span><span class="op">(</span></span>
<span>  geno <span class="op">=</span> <span class="va">exampleData</span><span class="op">$</span><span class="va">genotypes</span>,</span>
<span>  SNPinfo <span class="op">=</span> <span class="va">SNPs</span>,</span>
<span>  popName <span class="op">=</span> <span class="st">"Initial population"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">initPop</span><span class="op">)</span></span>
<span><span class="co">#&gt; Population: Initial population</span></span>
<span><span class="co">#&gt; Species: Statisticae exempli</span></span>
<span><span class="co">#&gt; Number of individuals: 100</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="traits-and-phenotyping-initialization">Traits and phenotyping initialization<a class="anchor" aria-label="anchor" href="#traits-and-phenotyping-initialization"></a>
</h3>
<p>In order to do some phenotyping simulation, we need to create a
trait.</p>
<p>We will use for that the example data
<code>exampleData$snpEffects</code> containing the “true” effects of
each QTN.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">exampleData</span><span class="op">$</span><span class="va">snpEffects</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span>
<span><span class="co">#&gt;   snp00006   snp00009   snp00011   snp00018   snp00026 </span></span>
<span><span class="co">#&gt; 0.05110930 0.08099233 0.18353259 0.02819434 0.14413695</span></span></code></pre></div>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">weight</span> <span class="op">&lt;-</span> <span class="va"><a href="../reference/trait.html">trait</a></span><span class="op">$</span><span class="fu">new</span><span class="op">(</span></span>
<span>  name <span class="op">=</span> <span class="st">"Weight"</span>,</span>
<span>  qtn <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">exampleData</span><span class="op">$</span><span class="va">snpEffects</span><span class="op">)</span>,</span>
<span>  qtnEff <span class="op">=</span> <span class="va">exampleData</span><span class="op">$</span><span class="va">snpEffects</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">weight</span><span class="op">)</span></span>
<span><span class="co">#&gt; trait: Weight</span></span>
<span><span class="co">#&gt; quantitative trait</span></span>
<span><span class="co">#&gt; number of QTN: 3333</span></span></code></pre></div>
<p>This trait object contain the genetic effects but in order to
simulate phenotyping experiments, we should also include some
environmental effects.</p>
<p>The environmental effect will be stored in the
<code>phenotyper</code> object. Here we create a phenotyper for which
the mean of the phenotypic value is equal to 100 and the initial
population have an heritability of 0.6, and the cost for phenotyping one
plot is 150.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">phenolab</span> <span class="op">&lt;-</span> <span class="va"><a href="../reference/phenotyper.html">phenotyper</a></span><span class="op">$</span><span class="fu">new</span><span class="op">(</span></span>
<span>  name <span class="op">=</span> <span class="st">"Pheno lab"</span>,</span>
<span>  traits <span class="op">=</span> <span class="va">weight</span>,</span>
<span>  plotCost <span class="op">=</span> <span class="fl">150</span>,</span>
<span>  mu <span class="op">=</span> <span class="fl">100</span>,</span>
<span>  he <span class="op">=</span> <span class="fl">0.6</span>,</span>
<span>  pop <span class="op">=</span> <span class="va">initPop</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">phenolab</span><span class="op">)</span></span>
<span><span class="co">#&gt; Phenotyper: Pheno lab</span></span>
<span><span class="co">#&gt; Traits: Weight</span></span>
<span><span class="co">#&gt; μ: 100</span></span>
<span><span class="co">#&gt; σ2: 25.6340237938445</span></span>
<span><span class="co">#&gt; Phenotyping cost (per plot): 150</span></span>
<span><span class="co">#&gt; timeEffect:</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="breeding-simulation">Breeding simulation<a class="anchor" aria-label="anchor" href="#breeding-simulation"></a>
</h2>
<p>Now the simulation is setup, we can write the simulation code using
the objects defined previously. For this example we will simulate a
simple breeding scheme consisting in phenotyping individuals and
selecting those with the highest phenotypic values and cross them
together.</p>
<div class="section level3">
<h3 id="phenotyping-individuals">phenotyping individuals<a class="anchor" aria-label="anchor" href="#phenotyping-individuals"></a>
</h3>
<p>First, let’s phenotype the initial population</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1809</span><span class="op">)</span> <span class="co"># set seed</span></span>
<span></span>
<span><span class="va">pheno</span> <span class="op">&lt;-</span> <span class="va">phenolab</span><span class="op">$</span><span class="fu">trial</span><span class="op">(</span>pop <span class="op">=</span> <span class="va">initPop</span>, rep <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span>
<span><span class="va">pheno</span><span class="op">$</span><span class="va">data</span></span>
<span><span class="co">#&gt;        ind    Weight rep phenotyper</span></span>
<span><span class="co">#&gt; 1 Coll0001  95.72423   1  Pheno lab</span></span>
<span><span class="co">#&gt; 2 Coll0001 102.72287   2  Pheno lab</span></span>
<span><span class="co">#&gt; 3 Coll0001  96.19940   3  Pheno lab</span></span>
<span><span class="co">#&gt; 4 Coll0002 105.09960   1  Pheno lab</span></span>
<span><span class="co">#&gt; 5 Coll0002 102.24307   2  Pheno lab</span></span>
<span><span class="co">#&gt; 6 Coll0002 102.28673   3  Pheno lab</span></span>
<span><span class="co">#&gt;  [ reached 'max' / getOption("max.print") -- omitted 294 rows ]</span></span>
<span><span class="va">pheno</span><span class="op">$</span><span class="va">cost</span></span>
<span><span class="co">#&gt; [1] 45000</span></span></code></pre></div>
<p>The <code>trial</code> method of the <code>phenotyper</code> class
return a list:</p>
<ul>
<li>
<code>data</code>: containing a <code>data.frame</code> of the
phenotypic data</li>
<li>
<code>cost</code>: a numeric value representing the cost of the
trial.</li>
</ul>
</div>
<div class="section level3">
<h3 id="selecting-individuals">Selecting individuals<a class="anchor" aria-label="anchor" href="#selecting-individuals"></a>
</h3>
<p>Here we will just select the individuals with the highest phenotypic
values.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># aggregate result by individual</span></span>
<span><span class="va">aggPheno</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/aggregate.html" class="external-link">aggregate</a></span><span class="op">(</span><span class="va">Weight</span> <span class="op">~</span> <span class="va">ind</span>, data <span class="op">=</span> <span class="va">pheno</span><span class="op">$</span><span class="va">data</span>, <span class="va">mean</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># sort result by "Weight" in ascending order</span></span>
<span><span class="va">aggPheno</span> <span class="op">&lt;-</span> <span class="va">aggPheno</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">aggPheno</span><span class="op">$</span><span class="va">Weight</span>, decreasing <span class="op">=</span> <span class="cn">T</span><span class="op">)</span>, <span class="op">]</span></span>
<span></span>
<span><span class="co"># set selection intensity</span></span>
<span><span class="va">i</span> <span class="op">&lt;-</span> <span class="fl">0.1</span></span>
<span></span>
<span><span class="co"># calculated number of selected individuals</span></span>
<span><span class="va">nSelected</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">floor</a></span><span class="op">(</span><span class="va">initPop</span><span class="op">$</span><span class="va">nInd</span> <span class="op">*</span> <span class="va">i</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># get names of selected individuals</span></span>
<span><span class="va">selectedInds</span> <span class="op">&lt;-</span> <span class="va">aggPheno</span><span class="op">$</span><span class="va">ind</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="va">nSelected</span><span class="op">]</span></span>
<span><span class="va">selectedInds</span></span>
<span><span class="co">#&gt;  [1] "Coll0068" "Coll0008" "Coll0074" "Coll0020" "Coll0045" "Coll0016"</span></span>
<span><span class="co">#&gt;  [7] "Coll0037" "Coll0036" "Coll0079" "Coll0046"</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="crossing-individuals">Crossing individuals<a class="anchor" aria-label="anchor" href="#crossing-individuals"></a>
</h3>
<p>In order to cross the individuals we need to create a crossing table
specifying:</p>
<ul>
<li>the first parent: <code>ind1</code>
</li>
<li>the second parent: <code>ind2</code>
</li>
<li>number of progenies for these two parents: <code>n</code>
</li>
<li>name of the progenies: <code>name</code>
</li>
</ul>
<p>In this example we will randomly cross the individuals together to
create 100 new individuals. The function <code>randomMate</code> can
generate such table.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">crossTable</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/randomMate.html">randomMate</a></span><span class="op">(</span><span class="va">selectedInds</span>, <span class="fl">100</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"gen-2_"</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">100</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">crossTable</span></span>
<span><span class="co">#&gt;       ind1     ind2 n   names</span></span>
<span><span class="co">#&gt; 1 Coll0079 Coll0046 1 gen-2_1</span></span>
<span><span class="co">#&gt; 2 Coll0037 Coll0016 1 gen-2_2</span></span>
<span><span class="co">#&gt; 3 Coll0045 Coll0068 1 gen-2_3</span></span>
<span><span class="co">#&gt; 4 Coll0036 Coll0074 1 gen-2_4</span></span>
<span><span class="co">#&gt; 5 Coll0045 Coll0074 1 gen-2_5</span></span>
<span><span class="co">#&gt; 6 Coll0068 Coll0008 1 gen-2_6</span></span>
<span><span class="co">#&gt;  [ reached 'max' / getOption("max.print") -- omitted 94 rows ]</span></span></code></pre></div>
<p>Now we have specified the crosses to do, the function
<code>makeCrosses</code> will create new individuals and return them as
a list:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">newIndsList</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/makeCrosses.html">makeCrosses</a></span><span class="op">(</span><span class="va">crossTable</span>, <span class="va">initPop</span><span class="op">)</span></span></code></pre></div>
<p>We can integrate these new individuals in a new population:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">newPop</span> <span class="op">&lt;-</span> <span class="va"><a href="../reference/population.html">population</a></span><span class="op">$</span><span class="fu">new</span><span class="op">(</span></span>
<span>  name <span class="op">=</span> <span class="st">"generation 2"</span>,</span>
<span>  inds <span class="op">=</span> <span class="va">newIndsList</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="automatisation">Automatisation<a class="anchor" aria-label="anchor" href="#automatisation"></a>
</h3>
<p>We have created one new generation, let’s create 3 more generations
other generation automatically integrating the code above in one
loop.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># create a list with all our populations</span></span>
<span><span class="va">nGen</span> <span class="op">&lt;-</span> <span class="fl">5</span> <span class="co"># total number of generation</span></span>
<span><span class="va">popList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/vector.html" class="external-link">vector</a></span><span class="op">(</span>mode <span class="op">=</span> <span class="st">"list"</span>, length <span class="op">=</span> <span class="va">nGen</span><span class="op">)</span></span>
<span><span class="va">popList</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">initPop</span></span>
<span><span class="va">popList</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">newPop</span></span>
<span></span>
<span><span class="va">nNew</span> <span class="op">&lt;-</span> <span class="fl">100</span> <span class="co"># number of new individual for each generation</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">gen</span> <span class="kw">in</span> <span class="fl">3</span><span class="op">:</span><span class="va">nGen</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">gen</span><span class="op">)</span></span>
<span>  <span class="co"># phenotyping</span></span>
<span>  <span class="va">pheno</span> <span class="op">&lt;-</span> <span class="va">phenolab</span><span class="op">$</span><span class="fu">trial</span><span class="op">(</span>pop <span class="op">=</span> <span class="va">popList</span><span class="op">[[</span><span class="va">gen</span> <span class="op">-</span> <span class="fl">1</span><span class="op">]</span><span class="op">]</span>, rep <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span>
<span></span>
<span>  <span class="co"># select individuals</span></span>
<span>  <span class="va">aggPheno</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/aggregate.html" class="external-link">aggregate</a></span><span class="op">(</span><span class="va">Weight</span> <span class="op">~</span> <span class="va">ind</span>, data <span class="op">=</span> <span class="va">pheno</span><span class="op">$</span><span class="va">data</span>, <span class="va">mean</span><span class="op">)</span></span>
<span>  <span class="va">aggPheno</span> <span class="op">&lt;-</span> <span class="va">aggPheno</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">aggPheno</span><span class="op">$</span><span class="va">Weight</span>, decreasing <span class="op">=</span> <span class="cn">T</span><span class="op">)</span>, <span class="op">]</span></span>
<span>  <span class="va">nSelected</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">floor</a></span><span class="op">(</span><span class="va">initPop</span><span class="op">$</span><span class="va">nInd</span> <span class="op">*</span> <span class="va">i</span><span class="op">)</span></span>
<span>  <span class="va">selectedInds</span> <span class="op">&lt;-</span> <span class="va">aggPheno</span><span class="op">$</span><span class="va">ind</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="va">nSelected</span><span class="op">]</span></span>
<span></span>
<span>  <span class="co"># cross individuals</span></span>
<span>  <span class="va">crossTable</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/randomMate.html">randomMate</a></span><span class="op">(</span></span>
<span>    inds <span class="op">=</span> <span class="va">selectedInds</span>,</span>
<span>    n <span class="op">=</span> <span class="va">nNew</span>,</span>
<span>    names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"gen-"</span>, <span class="va">gen</span>, <span class="st">"_"</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="va">nNew</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span>  <span class="va">newIndsList</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/makeCrosses.html">makeCrosses</a></span><span class="op">(</span><span class="va">crossTable</span>, <span class="va">popList</span><span class="op">[[</span><span class="va">gen</span> <span class="op">-</span> <span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span>  <span class="co"># create new population</span></span>
<span>  <span class="va">popList</span><span class="op">[[</span><span class="va">gen</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va"><a href="../reference/population.html">population</a></span><span class="op">$</span><span class="fu">new</span><span class="op">(</span></span>
<span>    name <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"generation"</span>, <span class="va">gen</span><span class="op">)</span>,</span>
<span>    inds <span class="op">=</span> <span class="va">newIndsList</span>,</span>
<span>    verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="co">#&gt; [1] 3</span></span>
<span><span class="co">#&gt; [1] 4</span></span>
<span><span class="co">#&gt; [1] 5</span></span>
<span><span class="va">popList</span></span>
<span><span class="co">#&gt; [[1]]</span></span>
<span><span class="co">#&gt; Population: Initial population</span></span>
<span><span class="co">#&gt; Species: Statisticae exempli</span></span>
<span><span class="co">#&gt; Number of individuals: 100</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[2]]</span></span>
<span><span class="co">#&gt; Population: generation 2</span></span>
<span><span class="co">#&gt; Species: Statisticae exempli</span></span>
<span><span class="co">#&gt; Number of individuals: 100</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[3]]</span></span>
<span><span class="co">#&gt; Population: generation 3</span></span>
<span><span class="co">#&gt; Species: Statisticae exempli</span></span>
<span><span class="co">#&gt; Number of individuals: 100</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[4]]</span></span>
<span><span class="co">#&gt; Population: generation 4</span></span>
<span><span class="co">#&gt; Species: Statisticae exempli</span></span>
<span><span class="co">#&gt; Number of individuals: 100</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[5]]</span></span>
<span><span class="co">#&gt; Population: generation 5</span></span>
<span><span class="co">#&gt; Species: Statisticae exempli</span></span>
<span><span class="co">#&gt; Number of individuals: 100</span></span></code></pre></div>
<p>Let’s look at the true genetic values of each population:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># create data.table with all genetic values</span></span>
<span><span class="va">GVs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">rbind</span>, <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">popList</span>, <span class="kw">function</span><span class="op">(</span><span class="va">pop</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>gvWeight <span class="op">=</span> <span class="va">weight</span><span class="op">$</span><span class="fu">gv</span><span class="op">(</span><span class="va">pop</span><span class="op">)</span>, pop <span class="op">=</span> <span class="va">pop</span><span class="op">$</span><span class="va">name</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># set pop variable as factor</span></span>
<span><span class="va">GVs</span><span class="op">$</span><span class="va">pop</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">GVs</span><span class="op">$</span><span class="va">pop</span>, levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">GVs</span><span class="op">$</span><span class="va">pop</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># create boxplot</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/boxplot.html" class="external-link">boxplot</a></span><span class="op">(</span><span class="va">gvWeight</span> <span class="op">~</span> <span class="va">pop</span>, data <span class="op">=</span> <span class="va">GVs</span><span class="op">)</span></span></code></pre></div>
<p><img src="example_files/figure-html/unnamed-chunk-12-1.png" width="768"></p>
</div>
</div>
<div class="section level2">
<h2 id="conclusion">Conclusion<a class="anchor" aria-label="anchor" href="#conclusion"></a>
</h2>
<p>We have build an algorithm for simulating a simple breeding camping.
In order to create more advance simulation, users can create their own
functions for selecting, mating individuals using the classes provided
by <code>breedSimulatR</code>. An exhaustive list of the fields and
methods available for each class can be found in the documentation of
the package and on <a href="https://ut-biomet.github.io/breedSimulatR/reference/index.html">the
package’s web site</a>.</p>
<div class="section level3">
<h3 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h3>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-PlantBreedGame" class="csl-entry">
Flutre, Timothée, Julien Diot, and Jacques David. 2019.
<span>“PlantBreedGame, A Serious Game That Puts Students in the
Breeder’s Seat.”</span> <em>Crop Science</em> 59 (October): 1374–75. <a href="https://doi.org/10.2135/cropsci2019.03.0183le" class="external-link">https://doi.org/10.2135/cropsci2019.03.0183le</a>.
</div>
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Julien Diot.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
